llvm/lib/Transforms/Scalar/DropUnnecessaryAssumes.cpp

```diff
@@ -17,13 +18,44 @@ using namespace llvm;
 using namespace llvm::PatternMatch;
 
 static bool affectedValuesAreEphemeral(ArrayRef<Value *> Affected) {
-  // If all the affected uses have only one use (part of the assume), then
-  // the assume does not provide useful information. Note that additional
-  // users may appear as a result of inlining and CSE, so we should only
-  // make this assumption late in the optimization pipeline.
-  // TODO: Handle dead cyclic usages.
-  // TODO: Handle multiple dead assumes on the same value.
-  return all_of(Affected, match_fn(m_OneUse(m_Value())));
+  // Check whether all the uses are ephemeral, i.e. recursively only used
+  // by assumes. In that case, the assume does not provide useful information.
+  // Note that additional users may appear as a result of inlining and CSE,
+  // so we should only make this assumption late in the optimization pipeline.
+  SmallSetVector<User *, 16> Worklist;
+  auto AddUser = [&](User *U) {
+    // Bail out if we need to inspect too many users.
+    if (Worklist.size() >= 32)
+      return false;
+    Worklist.insert(U);
+    return true;
+  };
+
+  for (Value *V : Affected)
+    for (User *U : V->users())
+      if (!AddUser(U))
+        return false;
```

Comment:
Hm, what if the Affected value isn't an Instruction? Wouldn't it be better to append the affected values to the Worklist, and let the code that follows dig the users?

Reply:
That would be simpler, but it wouldn't do the same thing. In particular, if the affected value itself is side-effecting, then we still want to treat it as ephemeral if it has no other uses. If we just pushed everything to the worklist, we'd treat them as non-ephemeral.

Reply:
Maybe add a quick comment saying that all affected values should be treated ephemeral?